src/engine.rs

```diff
@@ -507,11 +501,11 @@ impl Reedline {
     }
 
     fn set_history_navigation_based_on_line_buffer(&mut self) {
-        if self.line_buffer.is_empty()
-            || self.line_buffer.offset() != self.line_buffer.get_buffer().len()
-        {
-            self.history
-                .set_navigation(HistoryNavigationQuery::Normal(self.line_buffer.clone()));
+        if self.editor.is_empty() || self.editor.offset() != self.editor.get_buffer().len() {
+            self.history.set_navigation(HistoryNavigationQuery::Normal(
+                // Hack: Hight coupling point
+                self.editor.line_buffer.clone(),
```

Comment:
@sholderbach can you reiterate all the cases as to why we decided to use `LineBuffer` here? Let's see if we can find an alternative here

Reply:
This is for the bash style navigation, if you start on an empty line and go up through the history, and go back down, you end up on the empty line. If you started on a line with content it would restore that if the history is exhausted (moving beyond the most recent entry). In the current implementation, this behavior can be observed if you have content on the buffer but the cursor not at the end for prefix search. By storing it in the `HistoryNavigationQuery::Normal` Variant we won't loose track of that previous state. The navigation mode can only reset if you start editing the buffer.